enterprise/coordinator/history/configmapstore.go

```diff
@@ -0,0 +1,171 @@
+	}
+	cm.BinaryData[filepath.Base(key)] = value
+	_, err = s.client.CoreV1().ConfigMaps(s.namespace).Update(ctx, cm, metav1.UpdateOptions{})
+	return err
+}
+
+// CompareAndSwap updates the key to newVal if its current value is oldVal.
+func (s *ConfigMapStore) CompareAndSwap(key string, oldVal, newVal []byte) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	cm, err := s.client.CoreV1().ConfigMaps(s.namespace).Get(ctx, objectName(key), metav1.GetOptions{})
+	if err != nil && !(errors.IsNotFound(err) && len(oldVal) == 0) {
+		return err
+	}
+	// Treat non-existing config map as empty to allow initial set.
+	if errors.IsNotFound(err) {
+		cm = s.newEntry(key, newVal)
+		_, err := s.client.CoreV1().ConfigMaps(s.namespace).Create(ctx, cm, metav1.CreateOptions{})
+		return err
+	}
+	current, ok := cm.BinaryData[filepath.Base(key)]
+	if !ok {
+		return fmt.Errorf("key %q not found", key)
+	}
+	if !bytes.Equal(current, oldVal) {
+		return fmt.Errorf("object %q has changed since last read", key)
+	}
+	cm.BinaryData[filepath.Base(key)] = newVal
+	_, err = s.client.CoreV1().ConfigMaps(s.namespace).Update(ctx, cm, metav1.UpdateOptions{})
+	return err
+}
+
+// Watch watches for changes to the value of key.
+func (s *ConfigMapStore) Watch(key string) (<-chan []byte, func(), error) {
+	watcher, err := s.client.CoreV1().ConfigMaps(s.namespace).Watch(context.Background(), metav1.ListOptions{
+		LabelSelector: labels.Set(map[string]string{"app.kubernetes.io/managed-by": "contrast.edgeless.systems"}).AsSelector().String(),
+		FieldSelector: fields.OneTermEqualSelector("metadata.name", objectName(key)).String(),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	result := make(chan []byte)
+
+	go func() {
+		defer watcher.Stop()
+		defer close(result)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case event, ok := <-watcher.ResultChan():
+				if !ok {
+					return
+				}
+				switch event.Type {
+				case watch.Error:
+					return
+				case watch.Added, watch.Modified:
+					cm, ok := event.Object.(*corev1.ConfigMap)
+					if !ok {
+						continue
+					}
+					result <- cm.BinaryData[filepath.Base(key)]
+				}
+			}
+		}
+	}()
+
+	return result, cancel, nil
+}
+
+func (s *ConfigMapStore) newEntry(key string, value []byte) *corev1.ConfigMap {
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      objectName(key),
+			Namespace: s.namespace,
+			Labels:    map[string]string{"app.kubernetes.io/managed-by": "contrast.edgeless.systems"},
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "apps/v1",
+					Kind:       "StatefulSet",
+					Name:       "coordinator",
+					UID:        s.uid,
+				},
+			},
+		},
+		BinaryData: map[string][]byte{filepath.Base(key): value},
+	}
+}
+
+func objectName(key string) string {
+	return fmt.Sprintf("contrast-store-%s-%s", filepath.Dir(key), filepath.Base(key))
```

Comment:
This assumes that paths always have one level of nesting, and that there's no overlap between dir names (`foo/bar-baz` vs. `foo-bar/baz`,). These are probably fine assumptions, but we should at least document the constraints on the interface and enforce them here. "Keys must consist of two alphanumeric identifiers separated by forward slash".

Reply:
> Keys must consist of two alphanumeric identifiers separated by forward slash

... would be a way to avoid that (alphanumeric == `[a-zA-Z0-9]`).

Reply:
We should enforce this going forward.